oracle/oracle.go

```diff
@@ -0,0 +1,426 @@
+	for _, opt := range opts {
+		err := opt(oracle)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if oracle.phylum == nil {
+		err := withPhylum(config.GatewayEndpoint)(oracle)
+		if err != nil {
+			return nil, err
+		}
+	}
+	oracle.txConfigs = txConfigs()
+	traceOpts := []opttrace.Option{}
+	if config.OTLPEndpoint != "" {
+		traceOpts = append(traceOpts, opttrace.WithOTLPExporter(config.OTLPEndpoint))
+	}
+	t, err := opttrace.New(context.Background(), "oracle", traceOpts...)
+	if err != nil {
+		return nil, err
+	}
+	t.SetGlobalTracer()
+	oracle.tracer = t
+
+	oracle.log(context.Background()).WithFields(logrus.Fields{
+		"emulate_cc":       config.EmulateCC,
+		"phylum_path":      config.PhylumPath,
+		"gateway_endpoint": config.GatewayEndpoint,
+		"otlp_endpoint":    config.OTLPEndpoint,
+	}).Infof("new oracle")
+
+	return oracle, nil
+}
+
+func (orc *Oracle) log(ctx context.Context) *logrus.Entry {
+	return grpclogging.GetLogrusEntry(ctx, orc.logBase)
+}
+
+func txConfigs() func(context.Context, ...shiroclient.Config) []shiroclient.Config {
+	return func(ctx context.Context, extend ...shiroclient.Config) []shiroclient.Config {
+		fields := grpclogging.GetLogrusFields(ctx)
+		configs := []shiroclient.Config{
+			shiroclient.WithLogrusFields(fields),
+		}
+		if fields["req_id"] != nil {
+			logrus.WithField("req_id", fields["req_id"]).Infof("setting request id")
+			configs = append(configs, shiroclient.WithID(fmt.Sprint(fields["req_id"])))
+		}
+		configs = append(configs, extend...)
+		return configs
+	}
+}
+
+// setPhylumVersion sets the last seen phylum version and is concurrency safe.
+func (orc *Oracle) setPhylumVersion(version string) {
+	orc.phylumVersionMut.Lock()
+	defer orc.phylumVersionMut.Unlock()
+	orc.cachedPhylumVersion = version
+	if orc.cachedPhylumVersion != "" {
+		versionTotal.WithLabelValues(orc.serviceName, orc.version, orc.phylumServiceName, orc.cachedPhylumVersion).Inc()
+	}
+}
+
+// getLastPhylumVersion retrieves the last set phylum version and is concurrency safe.
+func (orc *Oracle) getLastPhylumVersion() string {
+	orc.phylumVersionMut.RLock()
+	defer orc.phylumVersionMut.RUnlock()
+	return orc.cachedPhylumVersion
+}
+
+func (orc *Oracle) phylumHealthCheck(ctx context.Context) []*healthcheck.HealthCheckReport {
+	sopts := orc.txConfigs(ctx)
+	ccHealth, err := orc.phylum.GetHealthCheck(ctx, []string{"phylum"}, sopts...)
+	if err != nil && !errors.Is(err, context.Canceled) {
+		return []*healthcheck.HealthCheckReport{{
+			ServiceName:    orc.phylumServiceName,
+			ServiceVersion: "",
+			Timestamp:      time.Now().Format(timestampFormat),
+			Status:         "DOWN",
+		}}
+	}
+	reports := ccHealth.GetReports()
+	for _, report := range reports {
+		if strings.EqualFold(report.GetServiceName(), orc.phylumServiceName) {
+			orc.setPhylumVersion(report.GetServiceVersion())
+			break
+		}
+	}
+	if orc.getLastPhylumVersion() == "" {
+		orc.log(ctx).Errorf("missing phylum version")
```

Comment:
Is this actually an error or more of a warning?

Reply:
the phylum version definitely needs to be specified, but perhaps this can be updated to default to `latest` if it isn't. What do you think?

Reply:
I guess it just seems odd to me to log an error but not return an error.  I seem to recall seeing this logged when the application is being set up which isn't really an error, but it was never an actionable log message.

Reply:
This function only returns health check reports at the moment, so there isn't really way to return an error. It calls setPhylumVersion right before looking up the cached version, so if it's empty then that would mean the health check did not report back a status for the phylum. Are you saying you're seeing this error under normal operations?

Reply:
I think this happens before init is called.  I recall seeing a similar error at some point in that situation.

Reply:
I see, well i've changed it from Error to warn